cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll

```diff
@@ -524,26 +525,48 @@ private predicate getFieldSizeOfClass(Class c, Type type, int size) {
   )
 }
 
-private predicate simpleOperandLocalFlowStep(Operand opFrom, Instruction iTo) {
-  // Certain dataflow steps (for instance `PostUpdateNode.getPreUpdateNode()`) generates flow to
-  // operands, so we include dataflow from those operands to the "result" of the instruction (i.e., to
-  // the instruction itself).
-  exists(PostUpdateNode post |
-    opFrom = post.getPreUpdateNode().asOperand() and
-    iTo.getAnOperand() = opFrom
+private predicate initializeParameterOfType(InitializeIndirectionInstruction init, Type type) {
+  init.getParameter().getType().getUnspecifiedType().(DerivedType).getBaseType() =
+    type.getUnspecifiedType()
+}
+
+private predicate isSingleFieldClass(Type type, Class cTo) {
+  exists(int size |
+    cTo.getSize() = size and
+    getFieldSizeOfClass(cTo, type, size)
+  )
+}
+
+private predicate simpleOperandLocalFlowStep(Instruction iFrom, Operand opTo) {
+  // Propagate flow from an instruction to its exact uses.
+  opTo.getDef() = iFrom
+  or
+  opTo = any(ReadSideEffectInstruction read).getSideEffectOperand() and
+  not iFrom.isResultConflated() and
+  iFrom = opTo.getAnyDef()
+  or
+  exists(InitializeIndirectionInstruction init |
+    iFrom = init and
+    opTo.(LoadOperand).getAnyDef() = init and
+    initializeParameterOfType(init, opTo.getType())
+  )
+  or
+  exists(LoadInstruction load |
+    load.getSourceValueOperand() = opTo and
+    opTo.getAnyDef() = iFrom and
+    isSingleFieldClass(iFrom.getResultType(), opTo.getType())
   )
 }
 
 cached
-private predicate simpleInstructionLocalFlowStep(Instruction iFrom, Instruction iTo) {
-  iTo.(CopyInstruction).getSourceValue() = iFrom
+private predicate simpleInstructionLocalFlowStep(Operand opFrom, Instruction iTo) {
+  iTo.(CopyInstruction).getSourceValueOperand() = opFrom
   or
-  iTo.(PhiInstruction).getAnOperand().getDef() = iFrom
+  iTo.(PhiInstruction).getAnInputOperand() = opFrom
   or
   // A read side effect is almost never exact since we don't know exactly how
   // much memory the callee will read.
-  iTo.(ReadSideEffectInstruction).getSideEffectOperand().getAnyDef() = iFrom and
-  not iFrom.isResultConflated()
+  iTo.(ReadSideEffectInstruction).getSideEffectOperand() = opFrom
```

Comment:
This rule is only needed until we change [`ArgumentNode`](https://github.com/github/codeql/blob/db45b2980620ce6968e8215ebb06291a15658529/cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll#L10) so the indirection argument becomes the _operand_ of the `ReadSideEffectInstruction` rather than the instruction itself. That can be in this PR or in a follow-up. One complication is that [the `toString` override](https://github.com/github/codeql/blob/db45b2980620ce6968e8215ebb06291a15658529/cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll#L425) will have to be moved to the operand also.

Reply:
Good point. I think I'll do this in a follow-up PR.